### PR TITLE
Params should render themselves

### PR DESCRIPTION
**Pull request by erlend-aasland**
Based off of python/cpython#107623

### CODE REVIEW

**Comment by erlend-aasland** [comment]
I think this is pretty neat, @AlexWaygood: fa29529c311696973b517ecd5f2c43cf613dd4e5

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -2640,6 +2640,15 @@ def get_displayname(self, i: int) -> str:
         else:
             return f'"argument {i}"'
 
+    def render_docstring(self) -> str:
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py:2643]: This may be a good case for `-> str | None`

Comment (erlend-aasland) [reply Tools/clinic/clinic.py:2643]: OTOH, I'm not sure it matters that much.

**Review by AlexWaygood — COMMENTED** [review]
This looks good, but I think we should go further. `format_docstring()` itself also looks like it should probably be a method on `Function` objects rather than a method on the `DSLParser`. It doesn't reference any state on the `DSLParser`, except for attributes on `self.function`. (This has implications for https://github.com/python/cpython/pull/107623!)

So, instead of `self.function.docstring = self.format_docstring()`, it would just be `self.function.format_docstring()` -- the `Function` object would know how to format its own docstring, instead of the `DSLParser` object reaching in and modifying the state of the `Function` object.

**Comment by erlend-aasland** [comment]
> This looks good, but I think we should go further. `format_docstring()` itself also looks like it should probably be a method on `Function` objects rather than a method on the `DSLParser`. It doesn't reference any state on the `DSLParser`, except for attributes on `self.function`. (This has implications for [python#107623](https://github.com/python/cpython/pull/107623)!)

Yes, that was my plan. I figured applying this change to Parameter was a nice start, since it makes for a small proof-of-concept PR.

> So, instead of `self.function.docstring = self.format_docstring()` in the `DSLParser.do_post_block_processing_cleanup()` method, it would just be `self.function.format_docstring()` -- the `Function` object would know how to format its own docstring, instead of the `DSLParser` object reaching in and modifying the state of the `Function` object.

I'm not sure yet how to refactor the generation phase. I have a proof-of-concept branch locally (I might have pushed it here, I don't remember, and I don't intend to find out before finishing writing this :) ) where I ended up with something like this in `render_function()`:

```python
    template_dict['docstring'] = f.render_docstring()
```

**Comment by erlend-aasland** [comment]
See gh-107790

**Pull request closed without merging**